### PR TITLE
Feature/bc 261 ability to update credentials

### PR DESCRIPTION
**Pull request by zendrias**
Sign in form was also merged into this branch from my computer, so this has the code for both...

During this sprint I...

- Created a **setting array where each item is a setting dropdown** in a separate module. We can easily add more setting dropdowns in the future using this method. (Keeps the code DRY too!) (Check it out here => [src/screens/AccountSettings/helper/data.ts](https://github.com/yeezick/bootcampr/compare/feature/BC-261-ability-to-update-credentials?expand=1#diff-407f3194a893d4d3cc4560ac35f220c29d8d11dda0f11bc28a77673a71c8050d))
- Added some poor styling lol
- Users can toggle whether to see their password while they are updating their credentials 
- The **authFormData** variable changes its initial state depending on which dropdown is open
- Only one dropdown can be open at a time, as per the Jira ticket requirements
- Button is disabled until the user fills out the form and addresses errors
*Note
I changed the following module to TypeScript, as it was giving me type errors.
[src/utilities/api/users.js → src/utilities/api/users.ts](https://github.com/yeezick/bootcampr/compare/feature/BC-261-ability-to-update-credentials?expand=1#diff-8491ed015b0166f70822a5f214cfebeb0dddb1771c143fa5e7fcdf60758b35b0)

### CODE REVIEW

File: src/screens/SignIn/SignIn.module.css [hunk 1]
```diff
@@ -0,0 +1,57 @@
+.sign_in_container {
```
Comment (HectorIlarraza) [inline src/screens/SignIn/SignIn.module.css:1]: ABC order

File: src/screens/AccountSettings/AccountSettings.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  return (
+    <div className={styles.account_settings_container}>
+
+      {settings.map(({ title, val, Component, props }) => (
+        <div key={val} className={styles.update_container}>
+          <div className={styles.setting_name}>
+            <p>Update {title}</p>
+
+            <button
+              onClick={() => handleToggleDropdown(val)}
```
Comment (HectorIlarraza) [inline src/screens/AccountSettings/AccountSettings.tsx:40]: is the anon function needed, just curious.

File: src/screens/AccountSettings/helper/FormErrors.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { FormErrorsProps } from "../../../utilities/types/AccountSettingsInterface";
+import styles from '../forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.module.css'
+
+const FormErrors = ({ VALIDATION_HELPERS, emailDropDownActive, authFormData }: FormErrorsProps) => {
+  const { emailsMatch, passwordsMatch, validEmail, emailFieldsFilledOut, passwordFieldsFilledOut } = VALIDATION_HELPERS
+
+  return (
+    <>
```
Comment (HectorIlarraza) [inline src/screens/AccountSettings/helper/FormErrors.tsx:8]: not sure if its ur prettier doing this but can you refactor this to make it easier to read the way its displayed now is kinda hard to follow without double checking

File: src/screens/AccountSettings/helper/UpdateFeedback.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
+
+
+interface UpdateCredentialsFeedbackMessageProps {
```
Comment (HectorIlarraza) [inline src/screens/AccountSettings/helper/UpdateFeedback.tsx:3]: can be moved into the "AccountSettingsInterface.tsx" file and imported here

**Review by HectorIlarraza — CHANGES_REQUESTED** [review]
Very impressive work Zena, had to really do some HW to see how you wrote all this. Just a couple of things to change and questions I'll need from you to approve.

File: package.json [hunk 5]
```diff
@@ -54,5 +53,8 @@
       "last 1 firefox version",
       "last 1 safari version"
     ]
+  },
+  "devDependencies": {
+    "node-sass": "^7.0.3"
```
Comment (yeezick) [inline package.json:58]: just curious, did we have to make it a dev dependency? 
will it be safe to build in production without this package?

File: src/App.tsx [hunk 6]
```diff
@@ -24,9 +25,12 @@ function App() {
           <Route path="/users/:id" element={<UserProfile />} />
           <Route path="/users/:id/account-setup" element={<RegisterUserInfo />} />
           <Route path="/users/:id/edit" element={<EditProfile />} />
-          <Route path="/projects/create" element={<CreateProject />} />
-          <Route path="/projects" element={<Projects />} />
-          <Route path="/projects/:id" element={<ProjectDetails />} />
+          < Route path="/projects">
+            <Route index element={<Projects />} />
```
[the remark this answers is not in the document]
Comment (yeezick) [reply src/App.tsx:29]: Yeah this is interesting! please feel free to mention it at stand up so we can all learn something

File: src/screens/AccountSettings/AccountSettings.module.css [hunk 7]
```diff
@@ -0,0 +1,51 @@
+* {
```
Comment (yeezick) [inline src/screens/AccountSettings/AccountSettings.module.css:1]: We have to discuss whether or not to change practices as a team before implementing them. 
If we choose not to go that route, you would lose time in refactoring

File: src/screens/AccountSettings/AccountSettings.module.css [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  margin: 0;
+  padding: 0;
+  box-sizing: border-box;
+}
+
+.account_settings_container {
```
Comment (yeezick) [inline src/screens/AccountSettings/AccountSettings.module.css:7]: same note here, all files currently use "class-name", if you would like to suggest a change to our style guide please bring it up to the team first.

File: src/screens/AccountSettings/AccountSettings.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
+import { useState } from 'react'
+import { initialDropdownState, settings } from './helper/data'
+import { DropDownSettings, AccountSettingsProps } from '../../utilities/types/AccountSettingsInterface'
+import styles from './AccountSettings.module.css'
+
```
Comment (yeezick) [inline src/screens/AccountSettings/AccountSettings.tsx:5]: whitespace

File: src/screens/AccountSettings/AccountSettings.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+const AccountSettings = ({ }: AccountSettingsProps): JSX.Element => {
```
Comment (yeezick) [inline src/screens/AccountSettings/AccountSettings.tsx:7]: if there are no expected props we can leave it as an empty `()`

File: src/screens/AccountSettings/AccountSettings.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    <div className={styles.account_settings_container}>
+
+      {settings.map(({ title, val, Component, props }) => (
+        <div key={val} className={styles.update_container}>
+          <div className={styles.setting_name}>
+            <p>Update {title}</p>
+
+            <button
+              onClick={() => handleToggleDropdown(val)}
+              className={`${dropdownModes[val as keyof DropDownSettings] ? styles.arrow : styles.active_arrow}`}>
```
Comment (yeezick) [inline src/screens/AccountSettings/AccountSettings.tsx:41]: i'm curious about the `as keyof` keyword, please teach us about it in stand up

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthFormInput.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import { useState } from 'react'
+import styles from './AuthSettingsFormDropdown.module.css'
+import { BsEyeFill, BsEyeSlash } from 'react-icons/bs';
+
+const AuthFormInput = ({ setAuthFormData, authFormData, field, type }: AuthFormInputProps): JSX.Element => {
+  // Constants
+  const PasswordInput = type.toLowerCase().includes('password')
+
+  // State Variables
+  const [showPassword, setShowPassword] = useState<boolean>(PasswordInput ? false : true)
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthFormInput.tsx:11]: if i'm reading this right, this would set "showPassword" to on the inital declaration of this state. 
let the user select when to toggle it manually by clicking a button. 
another cool feature i've seen is showing the user the password when they click on it and hiding it again when they unfocus

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthFormInput.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+
+  return (
+    <div className={styles.form_input_container}>
+
+      <label className={styles.auth_form_label} htmlFor={`${type}_input`}>{field}*</label>
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthFormInput.tsx:17]: you must be coding in python or something on the side lol, keep to camelCase. there aren't many use cases for dashes in JS

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthFormInput.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+          className={styles.toggle_hide_password}
+          onClick={() => setShowPassword(state => !state)}>
+          {showPassword ? < BsEyeFill className={styles.pwd_reveal} /> : <BsEyeSlash className={styles.pwd_reveal_gray} />}
+        </span>
+      )}
+    </div>
+  )
+}
+
+export default AuthFormInput;
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthFormInput.tsx:37]: no export defaults and also no newline at end of file

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+import { EmailFormData, PasswordFormData, AuthSettingsFormDropdownProps } from '../../../../utilities/types/AccountSettingsInterface'
+import styles from './AuthSettingsFormDropdown.module.css';
+import UpdateFeedback from '../../helper/UpdateFeedback'
+import FormErrors from '../../helper/FormErrors';
+
+const AuthSettingsFormDropdown = ({ fields, type }: AuthSettingsFormDropdownProps): JSX.Element => {
+  // Constants
+  const emailDropDownActive: boolean = type[1] === 'newEmail' // Lets us know whether or not the user is entering email or password info so we can change teh formData state accordingly.
+  const { id } = useParams()
+  const VALID_EMAIL_REGEX = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx:15]: this regex is thorough! add a comment here like "only allows ab@."

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+  // State Variables
+  const [authFormData, setAuthFormData] = useState<EmailFormData | PasswordFormData>(emailDropDownActive ? initialEmailFormData : initialPasswordFormData)
+  const [updateStatus, setUpdateStatus] = useState<string>("pending")
+  const [disableButton, setDisableButton] = useState<boolean>(true)
+
+  // Helper Functions
+  const VALIDATION_HELPERS = {
+    fetchAPI: async () => emailDropDownActive ? await updateUsersEmail(authFormData, id) : await updateUsersPassword(authFormData, id),
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx:24]: love this, clever. see if you can figure out how to avoid "any" type on "temp" 
temp is also not very semantic, not sure what it refers to

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (!passwordsMatch(temp) || fieldsNotFilledOut) setDisableButton(true)
+    else setDisableButton(false)
+  }
+
+  // Event Handlers
+  const handleUpdateCredentials = async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault()
+    const { fetchAPI } = VALIDATION_HELPERS
+
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx:53]: whitespace between consts

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    if (status === 201) return setUpdateStatus("authorized")
+    if (status === 401) return setUpdateStatus("unauthorized")
+
+    setUpdateStatus("error")
+  }
+
+
+  // Side Effects
+  useEffect(() => {
+    let temp: any = { ...authFormData }
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx:65]: same note from before, make temp descriptive and remove any type

File: src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        ))}
+
+        <button
+          disabled={disableButton}
+          type="submit"
+          className={styles.auth_form_submit_button}>
+          Update
+        </button>
+
+        < FormErrors VALIDATION_HELPERS={VALIDATION_HELPERS} emailDropDownActive={emailDropDownActive} authFormData={authFormData} />
```
Comment (yeezick) [inline src/screens/AccountSettings/forms/AuthSettingsFormDropdown/AuthSettingsFormDropdown.tsx:94]: fix your editor, it doesn't seem to be formatting correctly on this opening tag or for the new line at the end of this file

File: src/screens/AccountSettings/helper/FormErrors.tsx (hunk 3 above)
Comment (yeezick) [reply src/screens/AccountSettings/helper/FormErrors.tsx:8]: I agree, it also seems like we're repeating a lot of logic here, there might be a way to create some kind of 'error-generator' so that this file ages & scales better (if or when error cases grow)

File: src/screens/AccountSettings/helper/UpdateFeedback.tsx (hunk 4 above)
Comment (yeezick) [reply src/screens/AccountSettings/helper/UpdateFeedback.tsx:3]: ok so this makes me think of what would be best practice for our types
generally, if a variable is intended to be shared it should live within a shared file
in this case, it doesn't seem to be the case so would it be appropriate to live here or still worth relocating?
Next question would be whether or not it'd be worth to create a shared space (file or folder) to move our types into within the helper directory

File: src/screens/AccountSettings/helper/UpdateFeedback.tsx [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+  updateStatus: string;
+}
+
+const UpdateFeedback = ({ updateStatus }: UpdateCredentialsFeedbackMessageProps): JSX.Element => {
+
+  return (
+    <div>
+      {updateStatus === 'authorized' && <p>✔️ Update Successful</p>}
+      {updateStatus === 'unauthorized' && <p>❌ Wrong Password</p>}
```
Comment (yeezick) [inline src/screens/AccountSettings/helper/UpdateFeedback.tsx:12]: the answer for an 'error-generator' could be a switch statement
if a case does not have a "break" keyword i'm pretty sure it just cascades to the next case

File: src/screens/AccountSettings/helper/data.ts [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+// Prop Declarations
+const emailSettingsProps = {
+  type: ['password', 'newEmail', 'confirmNewEmail'],
+  fields: ["Password", 'New Email', "Confirm New Email"]
+}
+const passwordSettingsProps = {
+  type: ['password', 'newPassword', 'confirmNewPassword'],
+  fields: ["Current Password", 'New Password', "Confirm New Password"]
+}
+
```
Comment (yeezick) [inline src/screens/AccountSettings/helper/data.ts:13]: whitespace

Comment (yeezick) [reply src/screens/AccountSettings/helper/data.ts:13]: nice file structure

File: src/screens/SignIn/SignIn.module.css [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+  height: 35px;
+  padding: 5px;
+}
+
+.error_message {
+  color: red;
+  font-weight: bold;
+  font-size: 1.5rem;
+  letter-spacing: 4px;
+}
```
Comment (yeezick) [inline src/screens/SignIn/SignIn.module.css:57]: no new line

File: src/screens/SignIn/SignIn.tsx [hunk 23]
```diff
@@ -1,5 +1,91 @@
...
-export const SignIn: React.FC = () => {
-  return <div>SignIn</div>;
-};
\ No newline at end of file
+
+const SignIn: React.FC = (): JSX.Element => {
+  // State Variables
+  const [buttonDisabled, setButtonDisabled] = useState<boolean>(true)
+  const [invalidCredentials, setInvalidCredentials] = useState<boolean>(false)
+  const [formData, setFormData] = useState<SignInInterface>({ email: "", password: "" })
```
Comment (yeezick) [inline src/screens/SignIn/SignIn.tsx:15]: add the default state for formData to the data.ts file you created

File: src/utilities/types/AccountSettingsInterface.ts [hunk 24]
```diff
@@ -0,0 +1,52 @@
+export interface AccountSettingsProps {
+
```
Comment (yeezick) [inline src/utilities/types/AccountSettingsInterface.ts:2]: empty interface

File: src/utilities/types/AccountSettingsInterface.ts [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+  VALIDATION_HELPERS: {
+    fetchAPI: () => Promise<any> | { error: string; } |
+    {
+      error: {
+        status: number;
+        message: string;
+      }
+    };
+
+    emailsMatch: (temp: any | EmailFormData) => boolean;
```
Comment (yeezick) [inline src/utilities/types/AccountSettingsInterface.ts:37]: same note about temps and any

**Comment by yeezick** [comment]
@zendrias great work and good PR, a little too clever!! but always an opportunity to learn something in there. 
@HectorIlarraza great review as well my friend

Please be mindful the [sign-in PR ](https://github.com/yeezick/bootcampr/pull/6) is piggybacking on this one. make sure to merge and close sign in before merging this one. 

I trust Hector to finish the review on this PR, no further approval needed from me.

File: package.json (hunk 5 above)
Comment (zendrias) [reply package.json:58]: It wasn't working on my computer for some reason and I was following some stack overflow forum response to fix it. It had me do this, I meant to put it back before pushing, I just forgot...

File: src/screens/AccountSettings/AccountSettings.module.css (hunk 7 above)
Comment (zendrias) [reply src/screens/AccountSettings/AccountSettings.module.css:1]: I removed it

**Review by HectorIlarraza — CHANGES_REQUESTED** [review]
Incredible work here fixing everything that was requested Zena, Just one more suggestion on my part and I can approve you for the last phase.

File: src/screens/AccountSettings/helper/validation.ts [hunk 26]
```diff
@@ -0,0 +1,41 @@
+import { EmailFormData, PasswordFormData } from "../../../utilities/types/AccountSettingsInterface"
+import { updateUsersEmail, updateUsersPassword } from '../../../utilities/api/users'
+
+const VALID_EMAIL_REGEX = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/; // only allows ab@.
+interface FormDataCopy {
```
[the remark this answers is not in the document]
Comment (zendrias) [reply src/screens/AccountSettings/helper/validation.ts]: Done!

**Review by HectorIlarraza — APPROVED** [review]
Everything is all good Zena, approved!
